File: matiasmasso/Mat.NET/pwa.matiasmasso.es/DTO/Helpers/CryptoHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;

namespace DTO.Helpers
{
    public class CryptoHelper
    {
        //used to generate unique consumer ticket numbers
        public static string AsinFactory() => RandomString(10);

        public static string RandomString(int maxLen)
        {
            String range = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
            int rangeLength = range.Length;
            Random random = new Random();
            var sb = new StringBuilder();

            for (int i = 0; i < maxLen; i++)
            {
                int idx = random.Next(rangeLength);
                sb.Append(range.ElementAt(idx));
            }
            return sb.ToString();
        }

        //used by my custom implementation
        public static string Hash(string username, string password)
        {
            //salt generated on https://www.randomtextgenerator.com/
            var salt = "Si onda pero perico ocHo es cabe. Ola habia fue verde largo las ver quiso debia. Favoritas gas por ser redundara cincuenta. Tarda sento creia mayor monja sus ano. Un tropezando alumbraban proteccion conciertos tu id dilettante aficionado. Romantica esa eso orgullosa asi enamorado negarselo";
            using (var sha256 = SHA256.Create())
            {
                var saltedPassword = string.Format("{0}{1}{2}", salt, password, username);
                byte[] saltedPasswordAsBytes = Encoding.UTF8.GetBytes(saltedPassword);
                var retval = Convert.ToBase64String(sha256.ComputeHash(saltedPasswordAsBytes));
                return retval;
            }
        }

        //used by Microsoft identity tables
        public static string GetSha512Hash(string input)
        {
            using (var hashAlgorithm = SHA512.Create())
            {
                var byteValue = System.Text.Encoding.UTF8.GetBytes(input);
                var byteHash = hashAlgorithm.ComputeHash(byteValue);
                return Convert.ToBase64String(byteHash);
            }
        }


        //used by Redsys Tpv REST
        public static string GetSha256Hash(string input)
        {
            using (var hashAlgorithm = SHA256.Create())
            {
                var byteValue = System.Text.Encoding.UTF8.GetBytes(input);
                var byteHash = hashAlgorithm.ComputeHash(byteValue);
                return Convert.ToBase64String(byteHash);
            }
        }

        public static byte[] GetHMACSHA256(string msg, byte[] SignatureKey)
        {
            byte[] retval = null;
            try
            {
                // Obtain byte[] from input string
                byte[] msgBytes = System.Text.Encoding.UTF8.GetBytes(msg);

                // Initialize the keyed hash object.
                using (HMACSHA256 hmac = new HMACSHA256(SignatureKey))
                {

                    // Compute the hash of the input file.
                    retval = hmac.ComputeHash(msgBytes, 0, msgBytes.Length);
                }
            }
            catch (CryptographicException ex)
            {
                throw new CryptographicException(ex.Message);
            }
            return retval;
        }

        public static string HashMD5(string src)=> HashMD5(System.Text.ASCIIEncoding.ASCII.GetBytes(src));

        /// <summary>
        /// Media file signature
        /// </summary>
        /// <param name="oByteArray"></param>
        /// <returns></returns>
        public static string HashMD5(byte[] oByteArray)
        {
            //MD5CryptoServiceProvider Md5 = new System.Security.Cryptography.MD5CryptoServiceProvider(); //deprecated => MD5.Create
            var Md5 = MD5.Create();
            // Compute the hash value from the source
            byte[] ByteHash = Md5.ComputeHash(oByteArray);
            // And convert it to String format for return
            string retval = Convert.ToBase64String(ByteHash);
            return retval;
        }

        static public string EncodeTo64(string toEncode)
        {
            byte[] toEncodeAsBytes = System.Text.ASCIIEncoding.ASCII.GetBytes(toEncode);
            string returnValue = Convert.ToBase64String(toEncodeAsBytes);
            return returnValue;
        }

        public static Dictionary<string, string>? FromUrlFriendlyBase64Json(string? source)
        {
            Dictionary<string, string>? retval = null;
            if (source != null)
            {
                var sFromUrlFriendly = FromUrFriendlyBase64(source ?? "");
                var oBytes = System.Convert.FromBase64String(sFromUrlFriendly);
                if(oBytes != null)
                {
                    var sDecodedSource = System.Text.Encoding.UTF8.GetString(oBytes);
                    retval = Newtonsoft.Json.JsonConvert.DeserializeObject<Dictionary<string, string>>(sDecodedSource);
                }
            }
            return retval;
        }

        public static string FromUrFriendlyBase64(string source)
        {
            string retval = source.Replace("-", "+").Replace("_", "/").Replace("~", "=");
            return retval;
        }

        public static string UrlFriendlyBase64(string source)
        {
            string retval = "";
            if (!string.IsNullOrEmpty(source))
                retval = source.Replace("+", "-").Replace("/", "_").Replace("=", "~");
            return retval;
        }

        #region "Cifrado 3DES for TPV Redsys"

        //https://stackoverflow.com/questions/11413576/how-to-implement-triple-des-in-c-sharp-complete-example
        //https://pagosonline.redsys.es/conexion-rest.html

        public static byte[] Encrypt3DES(string plainText, byte[] SignatureKey)
        {
            byte[]? retval = null;

            if (string.IsNullOrEmpty(plainText))
                throw new FormatException("pedido vacio en Encrypt3DES");
            else
            {
                byte[] toEncryptArray = System.Text.Encoding.UTF8.GetBytes(plainText);
                var tdes = TripleDES.Create(); // new TripleDESCryptoServiceProvider();
                try
                {
                    // SALT used in 3DES encryptation process
                    byte[] SALT = new byte[] { 0, 0, 0, 0, 0, 0, 0, 0 };

                    // Block size 64 bit (8 bytes)
                    tdes.BlockSize = 64;

                    // Key Size 192 bit (24 bytes)
                    tdes.KeySize = 192;
                    tdes.Mode = CipherMode.CBC;
                    tdes.Padding = PaddingMode.Zeros;

                    tdes.IV = SALT;
                    tdes.Key = SignatureKey;

                    ICryptoTransform cTransform = tdes.CreateEncryptor();

                    // transform the specified region of bytes array to resultArray
                    retval = cTransform.TransformFinalBlock(toEncryptArray, 0, toEncryptArray.Length);

                    // Release resources held by TripleDes Encryptor
                    tdes.Clear();
                }
                catch (CryptographicException ex)
                {
                    throw new CryptographicException(ex.Message);
                }
            }

            return retval;
        }



        public static string TripleDesEncrypt(string key, string plainText)
        {
            var des = CreateDes(key);
            var ct = des.CreateEncryptor();
            var input = Encoding.UTF8.GetBytes(plainText);
            var output = ct.TransformFinalBlock(input, 0, input.Length);
            return Convert.ToBase64String(output);
        }

        public static string TripleDesDecrypt(string key, string cypherText)
        {
            var des = CreateDes(key);
            var ct = des.CreateDecryptor();
            var input = Convert.FromBase64String(cypherText);
            var output = ct.TransformFinalBlock(input, 0, input.Length);
            return Encoding.UTF8.GetString(output);
        }

        public static TripleDES CreateDes(string key)
        {
            MD5 md5 = MD5.Create(); // new MD5CryptoServiceProvider();
            TripleDES des = TripleDES.Create(); // new TripleDESCryptoServiceProvider();
            var desKey = md5.ComputeHash(Encoding.UTF8.GetBytes(key));
            des.Key = desKey;
            des.IV = new byte[des.BlockSize / 8];
            des.Padding = PaddingMode.PKCS7;
            des.Mode = CipherMode.ECB;
            return des;
        }

        #endregion



    }
}

File: matiasmasso/Mat.NET/FEB2/Helpers/CryptoHelper.vb
Imports System.Security.Cryptography

Public Class CryptoHelper

    Shared Function HashMD5(ByVal sCadena As String) As String
        Dim ueCodigo As New System.Text.UnicodeEncoding()
        Dim oByteArray As Byte() = ueCodigo.GetBytes(sCadena)
        Dim retval As String = HashMD5(oByteArray)
        Return retval
    End Function

    Shared Function HashMD5(oByteArray As Byte()) As String
        Dim Md5 As New System.Security.Cryptography.MD5CryptoServiceProvider()
        'Compute the hash value from the source
        Dim ByteHash() As Byte = Md5.ComputeHash(oByteArray)
        'And convert it to String format for return
        Dim retval As String = Convert.ToBase64String(ByteHash)
        Return retval
    End Function

    Shared Function FromUrFriendlyBase64(source As String) As String
        Dim retval As String = source.Replace("-", "+").Replace("_", "/").Replace("~", "=")
        Return retval
    End Function

    Shared Function UrlFriendlyBase64(source As String) As String
        Dim retval As String = ""
        If Not String.IsNullOrEmpty(source) Then
            retval = source.Replace("+", "-").Replace("/", "_").Replace("=", "~")
        End If
        Return retval
    End Function

    Shared Function UrlFriendlyBase64Json(oParameters As Dictionary(Of String, String)) As String
        Dim sJson As String = Newtonsoft.Json.JsonConvert.SerializeObject(oParameters)
        Dim oBytes() As Byte = System.Text.Encoding.UTF8.GetBytes(sJson)
        Dim sBase64 As String = System.Convert.ToBase64String(oBytes)
        Dim retval As String = UrlFriendlyBase64(sBase64)
        Return retval
    End Function

    Shared Function FromUrlFriendlyBase64Json(source As String) As Dictionary(Of String, String)
        Dim sFromUrlFriendly As String = FromUrFriendlyBase64(source)
        Dim oBytes() As Byte = System.Convert.FromBase64String(sFromUrlFriendly)
        Dim sDecodedSource As String = System.Text.Encoding.UTF8.GetString(oBytes)
        Dim o As Newtonsoft.Json.Linq.JObject = Newtonsoft.Json.JsonConvert.DeserializeObject(sDecodedSource)
        Dim retval As Dictionary(Of String, String) = o.ToObject(Of Dictionary(Of String, String))
        Return retval
    End Function



    Shared Function StringToHexadecimal(Src As String) As String
        Dim oByteArray() As Byte = System.Text.ASCIIEncoding.ASCII.GetBytes(Src)
        Dim sb As System.Text.StringBuilder = New System.Text.StringBuilder

        For i As Integer = 0 To oByteArray.Length - 1
            sb.Append(oByteArray(i).ToString("x"))
        Next

        Dim retval As String = sb.ToString()
        Return retval
    End Function

    Shared Function HexToString(ByVal Src As String) As String
        Dim sb As New System.Text.StringBuilder
        For i As Integer = 0 To Src.Length - 1 Step 2
            Dim SrcPair As String = Src.Substring(i, 2)
            Dim Hex As Integer = CInt("&H" & SrcPair)
            Dim sChar As String = ChrW(Hex)
            sb.Append(sChar)
        Next

        Dim retval As String = sb.ToString()
        Return retval
    End Function

    Shared Function HashString(inputString As String, hashName As String, exs As List(Of Exception)) As String
        Dim algorithm As Security.Cryptography.HashAlgorithm = Security.Cryptography.HashAlgorithm.Create(hashName)
        If algorithm Is Nothing Then
            exs.Add(New Exception("Unrecognized hash name " & hashName))
        End If
        Dim hash As Byte() = algorithm.ComputeHash(System.Text.Encoding.UTF8.GetBytes(inputString))
        Return Convert.ToBase64String(hash)
    End Function

    Shared Function GetSha256Hash(src As String) As String
        Dim oSha256 As Security.Cryptography.SHA256Managed = Security.Cryptography.SHA256Managed.Create()
        Dim hash As Byte() = oSha256.ComputeHash(System.Text.Encoding.UTF8.GetBytes(src))
        Dim retval As String = Convert.ToBase64String(hash)
        Return retval
    End Function

    Shared Function GetHMACSHA256(msg As String, SignatureKey As Byte()) As Byte()
        Dim retval As Byte() = Nothing
        Try
            'Obtain Byte() from input string
            Dim msgBytes As Byte() = System.Text.Encoding.UTF8.GetBytes(msg)

            'Initialize the keyed hash object.
            Using hmac As New HMACSHA256(SignatureKey)

                'Compute the hash of the input file.
                retval = hmac.ComputeHash(msgBytes, 0, msgBytes.Length)
            End Using
        Catch ex As CryptographicException
            Throw New CryptographicException(ex.Message)
        End Try
        Return retval
    End Function

    Shared Function Encrypt3DES(plainText As String, SignatureKey As Byte()) As Byte()
        Dim retval As Byte() = Nothing

        If String.IsNullOrEmpty(plainText) Then
            Throw New FormatException("pedido vacio en Encrypt3DES")
        Else
            Dim toEncryptArray As Byte() = System.Text.Encoding.UTF8.GetBytes(plainText)
            Dim tdes As New System.Security.Cryptography.TripleDESCryptoServiceProvider()
            Try
                ' SALT used in 3DES encryptation process
                Dim SALT As Byte() = {0, 0, 0, 0, 0, 0, 0, 0}

                ' Block size 64 bit (8 bytes)
                tdes.BlockSize = 64

                ' Key Size 192 bit (24 bytes)
                tdes.KeySize = 192
                tdes.Mode = CipherMode.CBC
                tdes.Padding = PaddingMode.Zeros

                tdes.IV = SALT
                tdes.Key = SignatureKey

                Dim cTransform As ICryptoTransform = tdes.CreateEncryptor()

                'transform the specified region of bytes array to resultArray
                retval = cTransform.TransformFinalBlock(toEncryptArray, 0, toEncryptArray.Length)

                'Release resources held by TripleDes Encryptor
                tdes.Clear()

            Catch ex As CryptographicException
                Throw New CryptographicException(ex.Message)
            End Try

        End If

        Return retval
    End Function


    Shared Function ToBase64(src As String) As String
        Dim oBytes As Byte() = System.Text.Encoding.UTF8.GetBytes(src)
        Dim retval As String = ToBase64(oBytes)
        Return retval
    End Function

    Shared Function ToBase64(oBytes As Byte()) As String
        Dim retval As String = Convert.ToBase64String(oBytes)
        Return retval
    End Function

    Shared Function FromBase64(src As String) As String
        Dim oBytes As Byte() = Convert.FromBase64String(src)
        Dim retval As String = System.Text.Encoding.UTF8.GetString(oBytes)
        Return retval
    End Function

End Class
